libbeat/filter/condition.go

```diff
@@ -11,247 +11,308 @@ import (
+	Int uint64
+	Str string
+}
+
 type Condition struct {
-	Equals   map[string]EqualsValue
-	Contains map[string]string
-	Regexp   map[string]*regexp.Regexp
-	Range    map[string]RangeValue
+	equals   map[string]EqualsValue
+	contains map[string]string
+	regexp   map[string]*regexp.Regexp
+	rangexp  map[string]RangeValue
+}
+
+func AvailableCondition(name string) bool {
+
+	switch name {
+	case "equals", "contains", "range", "regexp":
+		return true
+	default:
+		return false
+	}
 }
 
 func NewCondition(config ConditionConfig) (*Condition, error) {
 
 	c := Condition{}
-	c.Equals = map[string]EqualsValue{}
-	c.Contains = map[string]string{}
-	c.Regexp = map[string]*regexp.Regexp{}
-	c.Range = map[string]RangeValue{}
 
-	if err := c.AddEquals(config.Equals); err != nil {
-		return nil, err
-	}
-	if err := c.AddContains(config.Contains); err != nil {
-		return nil, err
-	}
-	if err := c.AddRegexp(config.Regexp); err != nil {
-		return nil, err
-	}
-	if err := c.AddRange(config.Range); err != nil {
-		return nil, err
+	if config.Equals != nil {
+		if err := c.setEquals(config.Equals); err != nil {
+			return nil, err
+		}
+	} else if config.Contains != nil {
+		if err := c.setContains(config.Contains); err != nil {
+			return nil, err
+		}
+	} else if config.Regexp != nil {
+		if err := c.setRegexp(config.Regexp); err != nil {
+			return nil, err
+		}
+	} else if config.Range != nil {
+		if err := c.setRange(config.Range); err != nil {
+			return nil, err
+		}
+	} else {
+		// empty condition
+		return nil, nil
 	}
 
 	return &c, nil
 }
 
-func (c *Condition) AddEquals(equals map[string]string) error {
+func (c *Condition) setEquals(cfg *ConditionFilter) error {
 
-	for field, value := range equals {
+	c.equals = map[string]EqualsValue{}
 
-		i, err := strconv.Atoi(value)
+	for field, value := range cfg.fields {
+		uint_value, err := extractInt(value)
 		if err == nil {
-			c.Equals[field] = EqualsValue{Int: i}
+			c.equals[field] = EqualsValue{Int: uint_value}
 		} else {
-			c.Equals[field] = EqualsValue{Str: value}
+			s_value, err := extractString(value)
+			if err != nil {
+				return err
+			}
+			c.equals[field] = EqualsValue{Str: s_value}
 		}
 	}
+
 	return nil
 }
 
-func (c *Condition) AddContains(contains map[string]string) error {
+func (c *Condition) setContains(cfg *ConditionFilter) error {
+
+	c.contains = map[string]string{}
+
+	for field, value := range cfg.fields {
+		switch v := value.(type) {
+		case string:
+			c.contains[field] = v
+		default:
+			return fmt.Errorf("unexpected type %T of %v", value, value)
+		}
+	}
 
-	c.Contains = contains
 	return nil
 }
 
-func (c *Condition) AddRegexp(r map[string]*regexp.Regexp) error {
-	for field, reg := range r {
-		c.Regexp[field] = reg
+func (c *Condition) setRegexp(cfg *ConditionFilter) error {
+
+	c.regexp = map[string]*regexp.Regexp{}
+	for field, value := range cfg.fields {
+		switch v := value.(type) {
+		case string:
+			c.regexp[field] = regexp.MustCompile(v)
```

Comment:
be carefull, `regexp.MustCompile` will panic on invalid regex. Better use `regexp.Compile`. I'd even recommend compiling the regex in Unpack method and return error on failure. ucfg will print additional context (which config failed and why) if Unpack fails.